### ISSUE TITLE
new cuda linking errors in downstream build systems

### ISSUE THREAD

**Issue by cyrush** [Alpine-DAV/ascent#807]
occurs in our `using-with-cmake` example post uberenv + spack update with newer versions of cmake (3.21.x?)

out of the box, `using-with-cmake` fails with a bunch of cuda linking errors.

`undefined reference to __cudaRegisterLinkedBinary`

and those in question these look like symbols that are from ascent, vtk-m, and vtk-h.

This is usually a symptom of `CUDA_SEPARABLE_COMPILATION`  being off, however we were using it globally in ascent and in the other libs. 

Just in case, adding the following in ascent:
`
    if(CUDA_FOUND)
        set_target_properties(ascent PROPERTIES CUDA_SEPARABLE_COMPILATION  ON)
        set_target_properties(ascent PROPERTIES CUDA_RESOLVE_DEVICE_SYMBOLS ON)
    endif()
`

Leads to a slew of `multiple definition of __cudaRegisterLinkedBinary` errors when compiling ascent's tests.  (tests compile and link when these are omitted)

This might be due to a new mix of what vtk-m pieces are compiled in vtk-h.

**Comment by mclarsen** [comment]
What happens if you just try to link `bcal_vtkm.o`?

**Comment by cyrush** [comment]
using-with-make compiles if I add` -ldl -lrt`  to the end (after libcudart_static.a)

```
root@0fd01c569b8c:/ascent/install-debug/examples/ascent/using-with-make# g++ -std=c++11 -I /uberenv_libs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/conduit-develop-weriip5helzlp4zbqltzknxambhwaeay/include/conduit -I /uberenv_libs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/conduit-develop-weriip5helzlp4zbqltzknxambhwaeay/include   -I/uberenv_libs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/hdf5-1.8.22-kpjet5b4cs2tpegzgfwyv4sgstue4zo2/include   -I /ascent/install-debug/include/ascent  ascent_render_example.cpp -Wl,-rpath,/ascent/install-debug/lib -Wl,-rpath,/uberenv_libs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/conduit-develop-weriip5helzlp4zbqltzknxambhwaeay/lib   -Wl,-rpath,/uberenv_libs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/hdf5-1.8.22-kpjet5b4cs2tpegzgfwyv4sgstue4zo2/lib       -Wl,-rpath,/uberenv_libs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/vtk-h-develop-clfkojd7jtv3eodd76hgtkfcfasrklqx/lib  -Wl,-rpath,/uberenv_libs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/vtk-m-1.6.0-5r2qyzyy7g7czn5k3cpmw7bgafm5jnlf/lib  /ascent/install-debug/lib/bcal_vtkm.o -L /ascent/install-debug/lib -lascent -lascent_flow -lascent_lodepng -lrover  -L /uberenv_libs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/vtk-h-develop-clfkojd7jtv3eodd76hgtkfcfasrklqx/lib -lvtkh_rendering -lvtkh_compositing -lvtkh_filters -lvtkh_core -lvtkh_utils -lvtkh_lodepng -L /uberenv_libs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/vtk-m-1.6.0-5r2qyzyy7g7czn5k3cpmw7bgafm5jnlf/lib -lvtkm_rendering-1.6 -lvtkm_filter_contour-1.6 -lvtkm_filter_gradient-1.6 -lvtkm_filter_extra-1.6 -lvtkm_filter_common-1.6 -lvtkm_worklet-1.6 -lvtkm_source-1.6 -lvtkm_io-1.6 -lvtkm_lodepng-1.6 -lvtkm_cont-1.6 -lvtkmdiympi_nompi -L /uberenv_libs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/conduit-develop-weriip5helzlp4zbqltzknxambhwaeay/lib -lconduit_blueprint -lconduit_relay -lconduit   -L /uberenv_libs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/hdf5-1.8.22-kpjet5b4cs2tpegzgfwyv4sgstue4zo2/lib -lhdf5 -lm -ldl /uberenv_libs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/zlib-1.2.11-rptnypqt45n7txo5aqqazlm7q6jobsxc/lib/libz.so  -ldl -lrt -lpthread     /usr/local/cuda/lib64/libcudadevrt.a /usr/local/cuda/lib64/libcudart_static.a -ldl -lrt -o ascent_render_example
```

That is a sign that bcal could work for cmake case, I'll try that next.

**Comment by cyrush** [comment]
bcal almost solves it, but i think maybe there are a few things sneaking by in the tpls or maybe a case we need to add for bcal creation?


```
[100%] Linking CXX executable ascent_render_example
/ascent/install-debug/lib/bcal_vtkm.o: In function `__cudaRegisterLinkedBinary_45_tmpxft_0000210c_00000000_6_Oscillator_cpp1_ii_1e80d97c':
link.stub:(.text+0x375a): undefined reference to `__fatbinwrap_45_tmpxft_0000210c_00000000_6_Oscillator_cpp1_ii_1e80d97c'
/ascent/install-debug/lib/bcal_vtkm.o: In function `__cudaRegisterLinkedBinary_41_tmpxft_0000211e_00000000_6_Source_cpp1_ii_959118ea':
link.stub:(.text+0x37aa): undefined reference to `__fatbinwrap_41_tmpxft_0000211e_00000000_6_Source_cpp1_ii_959118ea'
/ascent/install-debug/lib/bcal_vtkm.o: In function `__cudaRegisterLinkedBinary_41_tmpxft_00002117_00000000_6_Tangle_cpp1_ii_8dfec0d1':
link.stub:(.text+0x37fa): undefined reference to `__fatbinwrap_41_tmpxft_00002117_00000000_6_Tangle_cpp1_ii_8dfec0d1'
/ascent/install-debug/lib/bcal_vtkm.o: In function `__cudaRegisterLinkedBinary_42_tmpxft_0000211c_00000000_6_Wavelet_cpp1_ii_fef210d3':
link.stub:(.text+0x384a): undefined reference to `__fatbinwrap_42_tmpxft_0000211c_00000000_6_Wavelet_cpp1_ii_fef210d3'
collect2: error: ld returned 1 exit status
CMakeFiles/ascent_render_example.dir/build.make:127: recipe for target 'ascent_render_example' failed
make[2]: *** [ascent_render_example] Error 1
CMakeFiles/Makefile2:82: recipe for target 'CMakeFiles/ascent_render_example.dir/all' failed
make[1]: *** [CMakeFiles/ascent_render_example.dir/all] Error 2
Makefile:90: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by cyrush** [comment]
links perfectly fine with cmake 3.14 (w/o bcal) --  I'll spend some time to look a the difference in the link lines

**Comment by cyrush** [comment]
Got it, for newer cmake we need to explicitly set: 

`set_property(TARGET ascent_render_example PROPERTY CUDA_RESOLVE_DEVICE_SYMBOLS ON)`

**Comment by cyrush** [comment]
resolved in #790
